// @google/gemini-cli/dist/src/ui/components/Footer.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useState, useEffect, useRef } from 'react';
import { Box, Text } from 'ink';
import { theme } from '../semantic-colors.js';
import { shortenPath, tildeifyPath, getDisplayString, tokenLimit, } from '@google/gemini-cli-core';
import { ConsoleSummaryDisplay } from './ConsoleSummaryDisplay.js';
import process from 'node:process';
import { ThemedGradient } from './ThemedGradient.js';
import { MemoryUsageDisplay } from './MemoryUsageDisplay.js';
import { ContextUsageDisplay } from './ContextUsageDisplay.js';
import { DebugProfiler } from './DebugProfiler.js';
import { isDevelopment } from '../../utils/installationInfo.js';
import { useUIState } from '../contexts/UIStateContext.js';
import { useConfig } from '../contexts/ConfigContext.js';
import { useSettings } from '../contexts/SettingsContext.js';
import { useVimMode } from '../contexts/VimModeContext.js';

export const Footer = () => {
    const uiState = useUIState();
    const config = useConfig();
    const [quotaPercentage, setQuotaPercentage] = useState(null);
    const [resetTime, setResetTime] = useState(null);
    const [countdown, setCountdown] = useState("");

    // Helper function to fetch quota
    const fetchQuota = async (mountedRef) => {
        try {
            if (config.refreshUserQuota) {
                const q = await config.refreshUserQuota();
                if (mountedRef.current && q && q.buckets) {
                     // Try exact match first
                     let bucket = q.buckets.find(b => b.modelId === uiState.currentModel);
                     
                     // Fallback: Fuzzy search
                     if (!bucket) {
                        const normalizedModel = uiState.currentModel.toLowerCase().replace(/gemini-/, '').replace(/-preview/, '');
                        bucket = q.buckets.find(b => b.modelId && b.modelId.toLowerCase().includes(normalizedModel));
                     }

                     if (bucket) {
                         if (bucket.remainingFraction != null) {
                             setQuotaPercentage((bucket.remainingFraction * 100).toFixed(0));
                         }
                         if (bucket.resetTime) {
                             setResetTime(new Date(bucket.resetTime).getTime());
                         }
                     }
                }
            }
        } catch (e) {}
    };

    // Initial fetch on mount
    useEffect(() => {
        let mounted = { current: true };
        fetchQuota(mounted);
        return () => { mounted.current = false; };
    }, [uiState.currentModel]);

    // Debounced fetch on promptTokenCount change
    const { promptTokenCount } = uiState.sessionStats;
    useEffect(() => {
        if (!promptTokenCount || promptTokenCount < 100) return;

        const timer = setTimeout(() => {
            let mounted = { current: true };
            fetchQuota(mounted);
        }, 3000); // 3 seconds debounce

        return () => clearTimeout(timer);
    }, [promptTokenCount]);

    useEffect(() => {
        if (!resetTime) return;
        const timer = setInterval(() => {
            const now = Date.now();
            const diff = resetTime - now;
            if (diff <= 0) {
                setCountdown("00:00:00");
                clearInterval(timer);
                return;
            }
            const hours = Math.floor(diff / (1000 * 60 * 60));
            const minutes = Math.floor((diff % (1000 * 60 * 60)) / (1000 * 60));
            const seconds = Math.floor((diff % (1000 * 60)) / 1000);
            setCountdown(`${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`);
        }, 1000);
        return () => clearInterval(timer);
    }, [resetTime]);

    const settings = useSettings();
    const { vimEnabled, vimMode } = useVimMode();
    const { model, targetDir, debugMode, branchName, debugMessage, corgiMode, errorCount, showErrorDetails, nightly, isTrustedFolder, terminalWidth, } = {
        model: uiState.currentModel,
        targetDir: config.getTargetDir(),
        debugMode: config.getDebugMode(),
        branchName: uiState.branchName,
        debugMessage: uiState.debugMessage,
        corgiMode: uiState.corgiMode,
        errorCount: uiState.errorCount,
        showErrorDetails: uiState.showErrorDetails,
        nightly: uiState.nightly,
        isTrustedFolder: uiState.isTrustedFolder,
        terminalWidth: uiState.terminalWidth,
    };
    const showMemoryUsage = config.getDebugMode() || settings.merged.ui.showMemoryUsage;
    const hideCWD = settings.merged.ui.footer.hideCWD;
    const hideSandboxStatus = settings.merged.ui.footer.hideSandboxStatus;
    const hideModelInfo = settings.merged.ui.footer.hideModelInfo;
    const hideContextPercentage = settings.merged.ui.footer.hideContextPercentage;
    const pathLength = Math.max(20, Math.floor(terminalWidth * 0.25));
    const displayPath = shortenPath(tildeifyPath(targetDir), pathLength);
    const justifyContent = hideCWD && hideModelInfo ? 'center' : 'space-between';
    const displayVimMode = vimEnabled ? vimMode : undefined;
    const showDebugProfiler = debugMode || isDevelopment;

    // Dynamic RAM calculation
    const limit = tokenLimit(model) || 1048576;
    const ramPercentage = promptTokenCount ? (promptTokenCount / limit) * 100 : 0;
    
    // Request count extraction
    const modelStats = uiState.sessionStats.metrics.models[model];
    const requestCount = modelStats?.api?.totalRequests || 0;

    let ramColor = theme.text.secondary;
    let ramMessage = "";
    
    if (ramPercentage >= 80) {
        ramColor = theme.status.error;
        ramMessage = " - RISCHIO ALLUCINAZIONI, SALVA E APRI NUOVA CLI";
    } else if (ramPercentage >= 50) {
        ramColor = "#FFA500"; // Orange
    } else if (ramPercentage >= 20) {
        ramColor = theme.status.warning; // Yellow
    } else {
        ramColor = "green";
    }

    const budgetColor = quotaPercentage && quotaPercentage < 20 ? theme.status.error : (quotaPercentage && quotaPercentage < 50 ? theme.status.warning : "green");

    // EXPOSE DATA FOR EXTERNAL WIDGET (DEBUGGER HOOK)
    try {
        const targetGlobal = typeof globalThis !== 'undefined' ? globalThis : (typeof global !== 'undefined' ? global : {});
        targetGlobal.GEMINI_HUD_DATA = {
            budget: quotaPercentage !== null ? quotaPercentage : "Wait...",
            ram: ramPercentage ? ramPercentage.toFixed(1) : 0,
            reqs: requestCount || 0,
            timer: countdown || ""
        };
    } catch (e) { }

    return (_jsxs(Box, { justifyContent: justifyContent, width: terminalWidth, flexDirection: "row", alignItems: "center", paddingX: 1, children: [
        (showDebugProfiler || displayVimMode || !hideCWD) && (_jsxs(Box, { children: [
            showDebugProfiler && _jsx(DebugProfiler, {}), 
            displayVimMode && (_jsxs(Text, { color: theme.text.secondary, children: ["[", displayVimMode, "] "] })), 
            !hideCWD && (nightly ? (_jsxs(ThemedGradient, { children: [displayPath, branchName && _jsxs(Text, { children: [" (", branchName, "*)"] })] })) : (_jsxs(Text, { color: theme.text.link, children: [displayPath, branchName && (_jsxs(Text, { color: theme.text.secondary, children: [" (", branchName, "*)"] }))] }))), 
            debugMode && (_jsx(Text, { color: theme.status.error, children: ' ' + (debugMessage || '--debug') }))
        ] })), 
        
        !hideSandboxStatus && (_jsx(Box, { flexGrow: 1, alignItems: "center", justifyContent: "center", display: "flex", children: isTrustedFolder === false ? (_jsx(Text, { color: theme.status.warning, children: "untrusted" })) : process.env['SANDBOX'] && process.env['SANDBOX'] !== 'sandbox-exec' ? (_jsx(Text, { color: "green", children: process.env['SANDBOX'].replace(/^gemini-(?:cli-)?/, '') })) : process.env['SANDBOX'] === 'sandbox-exec' ? (_jsxs(Text, { color: theme.status.warning, children: ["macOS Seatbelt", ' ', _jsxs(Text, { color: theme.text.secondary, children: ["(", process.env['SEATBELT_PROFILE'], ")"] })] })) : (_jsxs(Text, { color: theme.status.error, children: ["no sandbox", terminalWidth >= 100 && (_jsx(Text, { color: theme.text.secondary, children: " (see /docs)" }))] })) })), 
        
        !hideModelInfo && (_jsxs(Box, { alignItems: "center", justifyContent: "flex-end", children: [
            _jsxs(Box, { alignItems: "center", children: [
                _jsxs(Text, { color: theme.text.accent, children: [
                    getDisplayString(model, config.getPreviewFeatures()), 
                    _jsx(Text, { color: theme.text.secondary, children: " /model " }), 
                    _jsxs(Text, { color: ramColor, children: [`RAM: ${ramPercentage.toFixed(1)}%${ramMessage}`] })
                ] }), 
                quotaPercentage && _jsxs(Text, { color: budgetColor, children: [` | Budget: ${quotaPercentage}%`, countdown && ` (Recharge: ${countdown})`, ` [Session Request: ${requestCount}]`] })
            ] }), 
            showMemoryUsage && _jsx(MemoryUsageDisplay, {})
        ] })), 
        
        _jsxs(Box, { alignItems: "center", children: [
            corgiMode && (_jsx(Box, { paddingLeft: 1, flexDirection: "row", children: _jsxs(Text, { children: [_jsx(Text, { color: theme.ui.symbol, children: "| " }), _jsx(Text, { color: theme.status.error, children: "\u25BC" }), _jsx(Text, { color: theme.text.primary, children: "(\u00B4" }), _jsx(Text, { color: theme.status.error, children: "\u1D25" }), _jsx(Text, { color: theme.text.primary, children: "`)" }), _jsx(Text, { color: theme.status.error, children: "\u25BC" })] }) })), 
            !showErrorDetails && errorCount > 0 && (_jsxs(Box, { paddingLeft: 1, flexDirection: "row", children: [_jsx(Text, { color: theme.ui.comment, children: "| " }), _jsx(ConsoleSummaryDisplay, { errorCount: errorCount })] }))
        ] })
    ] }));
};
